performance/multi-node/random_access.c

```diff
@@ -0,0 +1,1081 @@
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_cq_open", ret);
+		goto err_scq_open;
+	}
+
+	/* Open completion queue for recv completions */
+	ret = fi_cq_open(fabric->dom, &cq_attr, &p->rcq, NULL);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_cq_open", ret);
+		goto err_rcq_open;
+	}
+
+	return 0;
+
+	err_rcq_open: fi_close(&p->scq->fid);
+	p->scq = NULL;
+	err_scq_open: return ret;
+}
+
+static inline int bind_ep_res(fabric_t *fabric, int remote_rank)
+{
+	struct peer *p = &fabric->peers[remote_rank];
+	int ret;
+
+	/* Bind Send CQ with endpoint to collect send completions */
+	ret = fi_ep_bind(p->ep, &p->scq->fid, FI_TRANSMIT);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_ep_bind", ret);
+		return ret;
+	}
+
+	/* Bind Recv CQ with endpoint to collect recv completions */
+	ret = fi_ep_bind(p->ep, &p->rcq->fid, FI_RECV);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_ep_bind", ret);
+		return ret;
+	}
+
+	/* Bind AV with the endpoint to map addresses */
+	ret = fi_ep_bind(p->ep, &fabric->av->fid, 0);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_ep_bind", ret);
+		return ret;
+	}
+
+	ret = fi_enable(p->ep);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_enable", ret);
+		return ret;
+	}
+
+	return ret;
+}
+
+static inline int init_fabric(fabric_t *fabric)
+{
+	int ret;
+	uint64_t flags = 0;
+
+	/* Get fabric info */
+	ret = fi_getinfo(CT_FIVERSION, NULL, NULL, flags, fabric->hints,
+			&fabric->info);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_getinfo", ret);
+		return ret;
+	}
+
+	/* Open fabric */
+	ret = fi_fabric(fabric->info->fabric_attr, &fabric->fab, NULL);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_fabric", ret);
+		goto err_fabric_open;
+	}
+
+	/* Open domain */
+	ret = fi_domain(fabric->fab, fabric->info, &fabric->dom, NULL);
+	if (unlikely(ret)) {
+		ct_print_fi_error("fi_domain", ret);
+		goto err_domain_open;
+	}
+
+	return 0;
+
+	err_domain_open: fi_close(&fabric->fab->fid);
+	err_fabric_open: return ret;
+}
+
+static int init_endpoint(fabric_t *fabric, int remote_rank)
+{
+	struct peer *p = &fabric->peers[remote_rank];
+	int ret;
+	int i;
+
+	/* Open endpoint */
+	ret = fi_endpoint(fabric->dom, fabric->info, &p->ep, NULL);
+	if (ret) {
+		ct_print_fi_error("fi_endpoint", ret);
+		goto err_fi_endpoint;
+	}
+
+	/* Allocate endpoint resources */
+	ret = alloc_ep_res(fabric, remote_rank);
+	if (ret)
+		goto err_alloc_ep_res;
+
+	/* Bind EQs and AVs with endpoint */
+	ret = bind_ep_res(fabric, remote_rank);
+	if (ret)
+		goto err_bind_ep_res;
+
+	return 0;
+
+	err_bind_ep_res: free_ep_res(fabric, remote_rank);
+	err_alloc_ep_res: fi_close(&p->ep->fid);
+	err_fi_endpoint: return ret;
+}
+
+static int init_av(fabric_t *fabric)
+{
+	void *addr;
+	size_t addrlen = 0;
+	int ret;
+	int i;
+	struct peer *peers = fabric->peers;
+
+	/* use first ep since they will all have the same source addr */
+	for (i = 0; i < num_ranks; i++)
+		if (fabric->peers[i].ep)
+			break;
+
```

Comment:
I'm confused. 'i' will always be less than num ranks during the iteration of the for loop.